aii-pxelinux/src/main/perl/pxelinux.pm

```diff
@@ -335,155 +397,223 @@ sub pxe_ks_append
-        $this_app->debug (5, "Local booting from $path");
+        $self->debug (5, "Local booting from $path");
     } elsif ($cmd eq RESCUE || $cmd eq LIVECD || $cmd eq FIRMWARE) {
-        $path = link_filepath($cfg, $cmd);
-        if (! -f $path) {
-            my $fqdn = get_fqdn($cfg);
-            $this_app->error("Missing $cmd config file for $fqdn: $path");
-            return -1;
+        $path = $self->_link_filepath($cfg, $cmd, $variant);
+        if (! $self->file_exists($path) ) {
+            my $fqdn = $self->_get_fqdn($cfg);
+            $self->error("Missing $cmd config file for $fqdn: $path");
+            return 0;
         }
-        $this_app->debug (5, "Using $cmd from: $path");
+        $self->debug (5, "Using $cmd from: $path");
     } elsif ($cmd eq INSTALL) {
-        $path = filepath ($cfg);
-        $this_app->debug (5, "Installing on $path");
+        $path = $self->_filepath ($cfg, $variant);
+        $self->debug (5, "Installing on $path");
     } else {
-        $this_app->debug (5, "Unknown command");
-        return -1;
+        $self->debug (5, "Unknown command");
+        return 0;
     }
     # Set the same settings for every network interface that has a
     # defined IP address.
-    foreach my $st (values (%$t)) {
+    foreach my $st (values (%$interfaces)) {
         next unless $st->{ip};
-        my $dir = $this_app->option (NBPDIR);
-        my $lnname = "$dir/".hexip ($st->{ip});
+        my $dir = $self->_variant_option('nbpdir_opt', $variant);
+        my $lnname = "$dir/".$self->_hexip_filename ($st->{ip}, $variant);
         if ($cmd || ! -l $lnname) {
             if ($CAF::Object::NoAction) {
-                $this_app->info ("Would symlink $path to $lnname");
+                $self->info ("Would symlink $path to $lnname");
             } else {
-                unlink ($lnname);
-                # This must be stripped to work with chroot'edg
-                # environments.
+                my $unlink_status = $self->cleanup ($lnname);
+                if ( ! defined($unlink_status) ) {
+                    $self->error("Failed to delete $lnname (error=$self->{fail})");
+                } elsif ( $unlink_status == SUCCESS ) {
+                    $self->debug(1, "PXE link $lnname not found");
+                } else {
+                    $self->debug(1, "PXE link $lnname successfully removed");
+                };
+                # This must be stripped to work with chroot'ed environments.
                 $path =~ s{$dir/?}{};
                 symlink ($path, $lnname);
             }
         }
     }
-    return 0;
+
+    return 1;
+}
+
+
+# Wrapper function to call ksuserhooks() from aii-ks module.
+# The only role of this function is to ensure that ksuserhooks()
+# is always called the same way (in particular for NoAction
+# handling). Be sure to use it!
+sub _exec_userhooks {
+    my ($self, $cfg, $hook_path) = @_;
+
+    ksuserhooks ($cfg, $hook_path) unless $CAF::Object::NoAction;
```

Comment:
Because `ksuserhoots` is not `NoAction` safe, I'd say.... I created this method to execute the hook because it is called at several places and was called sometimes without any restriction, sometimes with this condition... I've the feeling that the good approach would be to have `ksuserhooks` safe with `NoAction` but until it is there it is probably a good idea to keep this condition, with the addition of a `TODO` mentioning that this should be removed when `ksuserhooks` is fixed. What do you think?